### PR TITLE
Improve weather and field tracking

### PR DESCRIPTION
**Pull request by Nailec**

### CODE REVIEW

**Comment by pmariglia** [comment]
Right off the bat a couple things stand out to me:

- this seems incomplete - the weather/field/trickroom counters arent decremented in the end of turn instructions
- Why have `-1` represent the weather/field/trickroom being inactive? Using `0` makes more sense as you can just check for a boolean on the number
- trickroom lasts 5 turns but your logic starts it at 4 turns (You put the integer 3 but I'm accounting for 0 being a number)

**Comment by Nailec** [comment]
> Right off the bat a couple things stand out to me:
> 
> * this seems incomplete - the weather/field/trickroom counters arent decremented in the end of turn instructions
> * Why have `-1` represent the weather/field/trickroom being inactive? Using `0` makes more sense as you can just check for a boolean on the number
> * trickroom lasts 5 turns but your logic starts it at 4 turns (You put the integer 3 but I'm accounting for 0 being a number)

- You are right I haven't done the end of turn decrement yet, I have to make sure of when it happens (weather stops before dealing damage but grassy terrain heals on its last turn).
- My thought was that on the last weather turn there's 0 remaining turn after but it's still active, which explains why I can't use 0 as a check for inactive. However starting the counter at 5 instead of 4 allows that but I should also rename remaining_turn into something else to avoid confusion
- You are right here too, I just googled "trick room duration" and it was written 4 turns in bold, however this didn't include the turn trick room was being set. I will fix that too

**Comment by pmariglia** [comment]
Just to keep consistent with how some other parts of the engine work I'd suggest having 0 represent "no longer active".

This is done for wish [here](https://github.com/pmariglia/showdown/blob/master/showdown/engine/instruction_generator.py#L872-L888).

Would that then make `remaining_turns` a bad name? I wouldn't think so. "0 remaining turns" to me implies that the effect is no longer active. The PokemonShowdown Client itself actually uses this as well.

File: showdown/engine/special_effects/abilities/on_switch_in.py [hunk 1]
```diff
@@ -6,7 +6,9 @@ def sandstream(state, attacking_side, attacking_pokemon, defending_side, defendi
         return [(
             constants.MUTATOR_WEATHER_START,
             constants.SAND,
-            state.weather
+            4,
+            state.weather,
+            state.remaining_weather_turns
         )]
     return None
```
Comment (pmariglia) [inline showdown/engine/special_effects/abilities/on_switch_in.py:14]: Each of these weather ability checks needs to now check for the existence of an extension item. If it exists the number would need to increase.

File: showdown/engine/instruction_generator.py [hunk 2]
```diff
@@ -99,11 +99,11 @@ def get_instructions_from_special_logic_move(mutator, attacking_pokemon, defendi
     mutator.apply(instructions.instructions)
     if move_name in weather_instructions and mutator.state.weather != move_name and mutator.state.weather not in constants.IRREVERSIBLE_WEATHER:
         new_instructions.append(
-            (constants.MUTATOR_WEATHER_START, move_name, mutator.state.weather)
+            (constants.MUTATOR_WEATHER_START, move_name, 4, mutator.state.weather, mutator.state.remaining_weather_turns)
```
Comment (pmariglia) [inline showdown/engine/instruction_generator.py:102]: similarly, extension items need to be checked here as well.

File: showdown/battle.py [hunk 3]
```diff
@@ -622,3 +623,36 @@ def __eq__(self, other):
...
+
+        total_turns = 5
+        if (
+                self.name == constants.RAIN and self.setter.item == "damprock" or
+                self.name == constants.SUN and self.setter.item == "heatrock" or
+                self.name == constants.HAIL and self.setter.item == "icyrock" or
+                self.name == constants.SAND and self.setter.item == "smoothrock" or
+                self.name in constants.TERRAINS and self.setter.item == "terrainextender"
+        ):
+            total_turns+=3
```
Comment (pmariglia) [inline showdown/battle.py:656]: This only checks the pokemon's item at the time of the call. The item could have been knocked off / tricked / etc. by the time this is called.

File: showdown/battle_modifier.py [hunk 4]
```diff
@@ -336,14 +336,21 @@ def cureteam(battle, split_msg):
 
 
 def weather(battle, split_msg):
+    weather_name = normalize_name(split_msg[2].split(':')[-1].strip())
+    logger.debug("Weather {} started".format(weather_name))
+
+    if weather_name == "none":
+        battle.weather = PersistentCondition()
+        return
+
```
Comment (pmariglia) [inline showdown/battle_modifier.py:345]: Need some tests to show the expected behaviour here - the PersistentCondition object isn't even imported in this module.

File: showdown/battle_modifier.py [hunk 5]
```diff
@@ -336,14 +336,21 @@ def cureteam(battle, split_msg):
...
         side = battle.opponent
     else:
         side = battle.user
 
-    weather_name = normalize_name(split_msg[2].split(':')[-1].strip())
-    logger.debug("Weather {} started".format(weather_name))
-    battle.weather = weather_name
+    battle.weather.name = weather_name
+    battle.weather.setter = side.active
+    battle.weather.elapsed_turns = 0
```
Comment (pmariglia) [inline showdown/battle_modifier.py:353]: shouldn't elapsed_turns be incremented somewhere (probably in [upkeep](https://github.com/pmariglia/showdown/blob/6c1602ab52b9c1c5d58de74a9722b804cd82c492/showdown/battle_modifier.py#L516) ) ?

**Comment by pmariglia** [comment]
There's a bit of examples in the documentation (or rather my pathetic attempt at documentation) in [ENGINE.md](https://github.com/pmariglia/showdown/blob/master/ENGINE.md) that needs to be updated because of this

**Comment by pmariglia** [comment]
Going to close this for now since it's been awhile with no activity.

Feel free to re-open if you'd like 🙂

**Pull request closed without merging**